src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,126 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        ensureCapacity();
+
+        if (index < size) {
+            System.arraycopy(elements, index, elements, index + 1, size - index);
+        }
+        elements[index] = value;
+        size++;
 
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
+    }
+
+    private int getIndex(T value) {
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(value, elements[i])) {
+                return i;
+            }
+        }
+        return -1;
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T currentElement = get(index);
+        System.arraycopy(elements, index + 1, elements, index, elements.length - (index + 1));
+        size--;
+        return currentElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndex(element);
+        if (index < 0) {
+            throw new NoSuchElementException("Element not found: " + element);
+        }
+
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private void ensureCapacity() {
+        if (size == elements.length) {
+            grow();
+        }
+    }
+
+    private void grow() {
+        int newCapacity = (int) (size * GROWTH_FACTOR);
```

Comment:
The growth factor should be applied to 'elements.length' instead of 'size' to ensure the array grows based on the current array capacity, not the number of elements.